src/Raven.Server/Utils/CpuUsageExtensionPoint.cs

```diff
@@ -0,0 +1,263 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Timers;
+using Raven.Client.Extensions;
+using Raven.Server.NotificationCenter.Notifications;
+using Raven.Server.NotificationCenter.Notifications.Details;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Sparrow.Logging;
+using Timer = System.Timers.Timer;
+
+namespace Raven.Server.Utils
+{
+    public class CpuUsageExtensionPoint : IDisposable
+    {
+        private readonly JsonContextPool _contextPool;
+        private readonly Logger _logger;
+        private readonly NotificationCenter.NotificationCenter _notificationCenter;
+        private readonly ProcessStartInfo _startInfo;
+        private readonly TimeSpan _timeout = TimeSpan.FromSeconds(10);
+
+        private Process _process;
+        private bool _didRestart;
+
+        private ExtensionPointRawData _data = new ExtensionPointRawData();
+        public ExtensionPointRawData BadData = new ExtensionPointRawData
+        {
+            ActiveCores = -1,
+            MachineCpuUsage = -1
+        };
+
+        public ExtensionPointRawData Data
+        {
+            get
+            {
+                if (IsDisposed && _logger.IsOperationsEnabled)
+                {
+                    _logger.Operations($"Try to get data from {nameof(CpuUsageExtensionPoint)} which was disposed");
+                }
+                return _data;
+            }
+        }
+
+        public bool IsDisposed { get; private set; }
+
+
+        public CpuUsageExtensionPoint(
+            JsonContextPool contextPool,
+            string exec,
+            string args,
+            Logger logger,
+            NotificationCenter.NotificationCenter notificationCenter)
+        {
+            _contextPool = contextPool;
+            _logger = logger;
+            _notificationCenter = notificationCenter;
+            _startInfo = new ProcessStartInfo
+            {
+                FileName = exec,
+                Arguments = args,
+                UseShellExecute = false,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                CreateNoWindow = true
+            };
+
+            InitNewProcessHandle();
+        }
+
+        public void Start()
+        {
+            try
+            {
+                _process.Start();
+            }
+            catch (Exception e)
+            {
+                NotifyWarning("Could not start cpu usage extension point process", e);
+
+                Dispose();
+            }
+
+            var _ = ReadProcess(); // explicitly starting async task without waiting for it
+            _ = ReadErrors();
+        }
+
+        private async Task ReadErrors()
+        {
+            var errors = await _process.StandardError.ReadLineAsync();
```

Comment:
This will read only one line of the error.

Reply:
Also need error handling here

Reply:
By "error handling" do you mean exception from `StreamReader` method?